csrc/multidevice/communicator.h

```diff
@@ -7,84 +14,107 @@
-
 namespace nvfuser {
 
 /*
    This file implements the class Communicator which sets up the inter-process
    Backend. This class contains inter-process information, such as the rank, the
    world size, as well as the Process Group that can be called to perform
-   inter-process communications
+   inter-process communications.
+
+   Each process is associated with a unique deviceId and device. The actual MPI
+   rank remains private to the class and should not be used by the user. The
+   communicator class holds privately the mappings ranks <-> device IDs <->
+   device.
 
-   Only one node configuration is supported for now.
-   TODO: extend to multinode.
 */
 
+using RankType = DeviceIdxType;
+
 // Supported backends. TODO: only tested with nccl for now
 enum class CommunicatorBackend { nccl, ucc, gloo };
 
-class Communicator {
+constexpr CommunicatorBackend comm_backend_default = CommunicatorBackend::nccl;
+constexpr int comm_server_local_rank_default = 0;
+constexpr int comm_master_port_default =
+    c10d::TCPStoreOptions::kDefaultPort; // 29500
+
+class TORCH_CUDA_CU_API Communicator {
  public:
   Communicator(
-      CommunicatorBackend backend = COMM_BACKEND_DEFAULT,
-      RankType server_local_rank = COMM_SERVER_LOCAL_RANK_DEFAULT);
+      CommunicatorBackend backend = comm_backend_default,
+      RankType server_local_rank = comm_server_local_rank_default);
+
+  Communicator(const Communicator&) = delete;
+  Communicator& operator=(const Communicator&) = delete;
 
   // returns if distributed config is available
   auto is_available() const {
     return is_available_;
   }
 
-  // returns the rank of the current process
-  auto rank() const {
-    return rank_;
-  }
-
   // returns the number of processes in the communicator
   auto size() const {
     return size_;
   }
 
-  // returns the local rank of the current process (within the node)
-  auto local_rank() const {
-    return local_rank_;
-  }
-
   // returns the local number of processes in the communicator (within the node)
   auto local_size() const {
     return local_size_;
   }
 
-  // returns the flattenend list of ranks of the communicator
-  auto ranks() const {
-    std::vector<RankType> ret(size());
-    std::iota(ret.begin(), ret.end(), 0);
-    return ret;
-  }
-
-  // performs a (blocking) send/receive p2p data transfer
-  void sendRecv(
-      RankType receiver_rank,
-      RankType sender_rank,
+  // performs a send/receive p2p data transfer
+  c10::intrusive_ptr<c10d::Work> sendRecv(
+      DeviceIdxType receiver,
+      DeviceIdxType sender,
       std::vector<at::Tensor>& tensor,
       int tag = 0);
 
   // performs a blocking barrier in the communicator
   void barrier() const {
-    pg_->barrier()->wait();
+    world_->barrier()->wait();
+  }
+
+  // returns the backend associated with a subset of devices/ranks given as
+  // arguments
+  c10::intrusive_ptr<c10d::Backend> getBackendForTeam(
+      const std::vector<DeviceIdxType>& devices);
```

Comment:
Should we do
```C++
using Team = std::vector<DeviceIdxType>;
c10::intrusive_ptr<c10d::Backend> getBackendForTeam(const Team& team);
```
for better readability?

Reply:
alright. I put `using Team = std::vector<DeviceIdxType>;` inside `multidevice.h` and replaced `std::vector<DeviceIdxType>` with `Team` everywhere